### PR TITLE
boundary conditions from python3 to hdf5

### PR DESCRIPTION
**Pull request by PhilipDeegan**
https://github.com/PHAREHUB/PHARE/issues/165
https://github.com/PHAREHUB/PHARE/issues/166

## Summary by CodeRabbit

- **New Features**
  - Added support for specifying boundary conditions for different simulation dimensions.
  - Enhanced the diagnostic tools to include boundary condition information in output files.

- **Bug Fixes**
  - Corrected loop declaration syntax for improved code consistency.
  - Improved handling of string data types in MPI utilities.

- **Documentation**
  - Updated documentation to reflect new methods and functionality related to boundary conditions.

- **Refactor**
  - Streamlined attribute writing in HDF5 files with new utility functions and error handling.
  - Refactored diagnostic model to include boundary conditions as part of patch properties.

- **Tests**
  - Implemented new tests to verify the correct setting and output of boundary conditions in diagnostics.

### CODE REVIEW

File: src/core/utilities/mpi_utils.hpp [hunk 1]
```diff
@@ -52,6 +52,8 @@ NO_DISCARD auto mpi_type_for()
         return MPI_UINT64_T;
     else if constexpr (std::is_same_v<char, Data>)
         return MPI_CHAR;
+    else if constexpr (std::is_same_v<std::string, Data>)
```
Comment (PhilipDeegan) [inline src/core/utilities/mpi_utils.hpp:55]: TORM I think

File: src/diagnostic/detail/h5writer.hpp [hunk 2]
```diff
@@ -133,19 +133,36 @@ class Writer
...
+            // runtime detection of unsupported types
+            if (unsupported)
+                throw std::runtime_error("Unsupported operation: Cannot write attribute: path("
+                                         + path + "), key(" + key + ")");
         });
     }
 
+    template<typename Dict>
+    static void writeGlobalAttributeDict(HighFiveFile& h5, Dict dict, std::string path)
+    {
```
Comment (PhilipDeegan) [inline src/diagnostic/detail/h5writer.hpp:154]: this one is for when all values are identical across mpi nodes, there is probably a better name

File: src/diagnostic/diagnostic_model_view.hpp [hunk 3]
```diff
@@ -37,7 +37,8 @@ class ModelView : public IModelView
     using GridLayout = typename Model::gridlayout_type;
     using PatchProperties
```
Comment (PhilipDeegan) [inline src/diagnostic/diagnostic_model_view.hpp:38]: it's possible we want to have a separate dict for patch properties, and for global properties
as now the patch properties has no need for "std::vector<string>", and the mpi attribute write function per mpi does not support it

**Comment by PhilipDeegan** [comment]
I figured it was a safe assumption that there wouldn't be different boundary conditions per level / model or solver type

**Comment by nicolasaunai** [comment]
caution about globals attributes considered in this PR that may be per rank from python

File: src/hdf5/detail/h5/h5_file.hpp [hunk 4]
```diff
@@ -141,9 +141,9 @@ class HighFiveFile
     template<typename Type, typename Size>
     void create_data_set_per_mpi(std::string const& path, Size const& dataSetSize)
     {
-        auto mpi_size = core::mpi::size();
-        auto sizes    = core::mpi::collect(dataSetSize, mpi_size);
-        auto paths    = core::mpi::collect(path, mpi_size);
+        auto const mpi_size = core::mpi::size();
```
Comment (nicolasaunai) [inline src/hdf5/detail/h5/h5_file.hpp:144]: #include "core/utilities/mpi_utils.hpp"

File: src/hdf5/detail/h5/h5_file.hpp [hunk 5]
```diff
@@ -154,13 +154,47 @@ class HighFiveFile
...
+        PHARE_DEBUG_DO(
+            auto paths = core::mpi::collect(keyPath, core::mpi::size());
+            for (auto const& path : paths)
+                PHARE_LOG_LINE_STR(std::to_string(core::mpi::size()) << " " << path)
+            if (!core::all(paths, [&](auto const& path) { return path == paths[0]; }))
+                throw std::runtime_error("Function does not support different paths per mpi core");
+        )
+        // clang-format on
+
+        constexpr bool data_is_vector = core::is_std_vector_v<Data>;
```
Comment (nicolasaunai) [inline src/hdf5/detail/h5/h5_file.hpp:174]: #include "core/utilities/types.hpp"
#include "core/utilities/meta/meta_utilities.hpp" (for is_iterable_v not in this diff)

**Review by nicolasaunai — APPROVED** [review]
few comments here and there

File: src/diagnostic/detail/h5writer.hpp (hunk 2 above)
Comment (PhilipDeegan) [reply src/diagnostic/detail/h5writer.hpp:155]: > rename the MPI one?

they both use MPI :\

Comment (nicolasaunai) [reply src/diagnostic/detail/h5writer.hpp:155]: the first calls `per_mpi` the other not, I don't understand

Comment (PhilipDeegan) [reply src/diagnostic/detail/h5writer.hpp:155]: they both do collective calls, just the `global` (or not `per_mpi`) version expects all values to be identical

**Review by nicolasaunai — APPROVED** [review]
